From 13867367c67abd9d207aef97a001842899f36b30 Mon Sep 17 00:00:00 2001
From: Jeff Peeler <jpeeler@redhat.com>
Date: Tue, 2 Apr 2013 18:08:48 -0400
Subject: [PATCH] Adjust to handle parallel installed packages

---
 heat/__init__.py |   33 +++++++++++++++++++++++++++++++++
 1 files changed, 33 insertions(+), 0 deletions(-)

diff --git a/heat/__init__.py b/heat/__init__.py
index 57e2d0f..d9df4ce 100644
--- a/heat/__init__.py
+++ b/heat/__init__.py
@@ -19,3 +19,36 @@ from heat.openstack.common import gettextutils
 
 
 gettextutils.install('heat')
+
+import sys
+import pkg_resources
+
+# If there is a conflicting non egg module,
+# i.e. an older standard system module installed,
+# then replace it with this requirement
+def replace_dist(requirement):
+    try:
+       return pkg_resources.require(requirement)
+    except pkg_resources.VersionConflict:
+        e = sys.exc_info()[1]
+        dist = e.args[0]
+        req = e.args[1]
+        if dist.key == req.key and not dist.location.endswith('.egg'):
+            del pkg_resources.working_set.by_key[dist.key]
+            # We assume there is no need to adjust sys.path
+            # and the associated pkg_resources.working_set.entries
+            try:
+                return pkg_resources.require(requirement)
+            except pkg_resources.DistributionNotFound:
+                print 'Not found, skipping', requirement
+                return {}
+
+replace_dist("WebOb >= 1.0")
+replace_dist("SQLAlchemy >= 0.6.3")
+replace_dist("Routes >= 1.12.3")
+replace_dist("PasteDeploy >= 1.5")
+# This hack is needed because replace_dist() results in
+# the standard paste module path being at the start of __path__.
+# TODO: See can we get pkg_resources to do the right thing directly
+import paste
+paste.__path__.insert(0, paste.__path__.pop(-1))
